edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/ServiceProxy.cs

```diff
@@ -175,6 +177,60 @@ public static void BadRequestResult(string id, HttpStatusCode statusCode)
             }
         }
 
+        /// <summary>
+        /// This is a helper class used to pull down the devices/modules in scope for every nested Edge Device.
+        /// It maintains a queue of ServiceIdentitiesIterator, each representating a child Edge device
+        /// in the subtree, and uses these iterators to perform breadth-first-expansion on the tree.
+        /// </summary>
+        class NestedServiceIdentitiesIterator : IServiceIdentitiesIterator
+        {
+            IDeviceScopeApiClient rootClient;
+            Queue<IServiceIdentitiesIterator> remainingEdgeNodes;
+
+            public NestedServiceIdentitiesIterator(IDeviceScopeApiClient securityScopesApiClient)
+            {
+                this.remainingEdgeNodes = new Queue<IServiceIdentitiesIterator>();
+
+                this.rootClient = Preconditions.CheckNotNull(securityScopesApiClient);
+                this.remainingEdgeNodes.Enqueue(new ServiceIdentitiesIterator(securityScopesApiClient));
+            }
+
+            public bool HasNext => this.remainingEdgeNodes.Count > 0;
+
+            public Task<IEnumerable<ServiceIdentity>> GetNext()
+            {
+                // Check for the empty-case
+                if (this.remainingEdgeNodes.Count == 0)
+                {
+                    return Task.FromResult(Enumerable.Empty<ServiceIdentity>());
+                }
+
+                // Get the head without popping, as each iterator could potentially require
+                // multiple calls to GetNext() to fully drain, we can't remove it from
+                // the queue until it's fully completed
+                IServiceIdentitiesIterator currentNode = this.remainingEdgeNodes.Peek();
+                IList<ServiceIdentity> results = currentNode.GetNext().Result.ToList();
```

Comment:
q: why blocking call? Do we support C# 8, and can we implement IAsyncEnumerator?

Reply:
I wanted to use AsyncEnumerator when I coded this.. it was not available at that time! :)